File: examples/go/babybear-ext4/BabyBearExt4Demo.runar.go
package contract

import runar "github.com/icellan/runar/packages/runar-go"

// BabyBearExt4Demo demonstrates Baby Bear Ext4 (quartic extension field)
// arithmetic and the core FRI colinearity folding relation used in SP1 STARK
// proofs.
type BabyBearExt4Demo struct {
	runar.SmartContract
}

// CheckMul verifies Ext4 multiplication (all 4 components).
func (bb *BabyBearExt4Demo) CheckMul(
	a0, a1, a2, a3,
	b0, b1, b2, b3,
	e0, e1, e2, e3 runar.Bigint,
) {
	runar.Assert(runar.BbExt4Mul0(a0, a1, a2, a3, b0, b1, b2, b3) == e0)
	runar.Assert(runar.BbExt4Mul1(a0, a1, a2, a3, b0, b1, b2, b3) == e1)
	runar.Assert(runar.BbExt4Mul2(a0, a1, a2, a3, b0, b1, b2, b3) == e2)
	runar.Assert(runar.BbExt4Mul3(a0, a1, a2, a3, b0, b1, b2, b3) == e3)
}

// CheckInv verifies Ext4 inverse (all 4 components).
func (bb *BabyBearExt4Demo) CheckInv(
	a0, a1, a2, a3,
	e0, e1, e2, e3 runar.Bigint,
) {
	runar.Assert(runar.BbExt4Inv0(a0, a1, a2, a3) == e0)
	runar.Assert(runar.BbExt4Inv1(a0, a1, a2, a3) == e1)
	runar.Assert(runar.BbExt4Inv2(a0, a1, a2, a3) == e2)
	runar.Assert(runar.BbExt4Inv3(a0, a1, a2, a3) == e3)
}

// CheckFriFold verifies the core FRI colinearity folding relation.
func (bb *BabyBearExt4Demo) CheckFriFold(
	x,
	fx0, fx1, fx2, fx3,
	fnx0, fnx1, fnx2, fnx3,
	a0, a1, a2, a3,
	eg0, eg1, eg2, eg3 runar.Bigint,
) {
	s0 := runar.BbFieldAdd(fx0, fnx0)
	s1 := runar.BbFieldAdd(fx1, fnx1)
	s2 := runar.BbFieldAdd(fx2, fnx2)
	s3 := runar.BbFieldAdd(fx3, fnx3)
	inv2 := runar.BbFieldInv(2)
	hs0 := runar.BbFieldMul(s0, inv2)
	hs1 := runar.BbFieldMul(s1, inv2)
	hs2 := runar.BbFieldMul(s2, inv2)
	hs3 := runar.BbFieldMul(s3, inv2)
	d0 := runar.BbFieldSub(fx0, fnx0)
	d1 := runar.BbFieldSub(fx1, fnx1)
	d2 := runar.BbFieldSub(fx2, fnx2)
	d3 := runar.BbFieldSub(fx3, fnx3)
	ad0 := runar.BbExt4Mul0(a0, a1, a2, a3, d0, d1, d2, d3)
	ad1 := runar.BbExt4Mul1(a0, a1, a2, a3, d0, d1, d2, d3)
	ad2 := runar.BbExt4Mul2(a0, a1, a2, a3, d0, d1, d2, d3)
	ad3 := runar.BbExt4Mul3(a0, a1, a2, a3, d0, d1, d2, d3)
	inv2x := runar.BbFieldInv(runar.BbFieldMul(2, x))
	at0 := runar.BbFieldMul(ad0, inv2x)
	at1 := runar.BbFieldMul(ad1, inv2x)
	at2 := runar.BbFieldMul(ad2, inv2x)
	at3 := runar.BbFieldMul(ad3, inv2x)
	g0 := runar.BbFieldAdd(hs0, at0)
	g1 := runar.BbFieldAdd(hs1, at1)
	g2 := runar.BbFieldAdd(hs2, at2)
	g3 := runar.BbFieldAdd(hs3, at3)
	runar.Assert(g0 == eg0)
	runar.Assert(g1 == eg1)
	runar.Assert(g2 == eg2)
	runar.Assert(g3 == eg3)
}

File: examples/zig/babybear-ext4/BabyBearExt4Demo.runar.zig
const runar = @import("runar");

pub const BabyBearExt4Demo = struct {
    pub const Contract = runar.SmartContract;

    pub fn init() BabyBearExt4Demo {
        return .{};
    }

    pub fn checkMul(
        self: *const BabyBearExt4Demo,
        a0: i64, a1: i64, a2: i64, a3: i64,
        b0: i64, b1: i64, b2: i64, b3: i64,
        e0: i64, e1: i64, e2: i64, e3: i64,
    ) void {
        _ = self;
        runar.assert(runar.bbExt4Mul0(a0, a1, a2, a3, b0, b1, b2, b3) == e0);
        runar.assert(runar.bbExt4Mul1(a0, a1, a2, a3, b0, b1, b2, b3) == e1);
        runar.assert(runar.bbExt4Mul2(a0, a1, a2, a3, b0, b1, b2, b3) == e2);
        runar.assert(runar.bbExt4Mul3(a0, a1, a2, a3, b0, b1, b2, b3) == e3);
    }

    pub fn checkInv(
        self: *const BabyBearExt4Demo,
        a0: i64, a1: i64, a2: i64, a3: i64,
        e0: i64, e1: i64, e2: i64, e3: i64,
    ) void {
        _ = self;
        runar.assert(runar.bbExt4Inv0(a0, a1, a2, a3) == e0);
        runar.assert(runar.bbExt4Inv1(a0, a1, a2, a3) == e1);
        runar.assert(runar.bbExt4Inv2(a0, a1, a2, a3) == e2);
        runar.assert(runar.bbExt4Inv3(a0, a1, a2, a3) == e3);
    }

    pub fn checkFriFold(
        self: *const BabyBearExt4Demo,
        x: i64,
        fx0: i64, fx1: i64, fx2: i64, fx3: i64,
        fnx0: i64, fnx1: i64, fnx2: i64, fnx3: i64,
        a0: i64, a1: i64, a2: i64, a3: i64,
        eg0: i64, eg1: i64, eg2: i64, eg3: i64,
    ) void {
        _ = self;
        const s0 = runar.bbFieldAdd(fx0, fnx0);
        const s1 = runar.bbFieldAdd(fx1, fnx1);
        const s2 = runar.bbFieldAdd(fx2, fnx2);
        const s3 = runar.bbFieldAdd(fx3, fnx3);
        const inv2 = runar.bbFieldInv(2);
        const hs0 = runar.bbFieldMul(s0, inv2);
        const hs1 = runar.bbFieldMul(s1, inv2);
        const hs2 = runar.bbFieldMul(s2, inv2);
        const hs3 = runar.bbFieldMul(s3, inv2);
        const d0 = runar.bbFieldSub(fx0, fnx0);
        const d1 = runar.bbFieldSub(fx1, fnx1);
        const d2 = runar.bbFieldSub(fx2, fnx2);
        const d3 = runar.bbFieldSub(fx3, fnx3);
        const ad0 = runar.bbExt4Mul0(a0, a1, a2, a3, d0, d1, d2, d3);
        const ad1 = runar.bbExt4Mul1(a0, a1, a2, a3, d0, d1, d2, d3);
        const ad2 = runar.bbExt4Mul2(a0, a1, a2, a3, d0, d1, d2, d3);
        const ad3 = runar.bbExt4Mul3(a0, a1, a2, a3, d0, d1, d2, d3);
        const inv2x = runar.bbFieldInv(runar.bbFieldMul(2, x));
        const at0 = runar.bbFieldMul(ad0, inv2x);
        const at1 = runar.bbFieldMul(ad1, inv2x);
        const at2 = runar.bbFieldMul(ad2, inv2x);
        const at3 = runar.bbFieldMul(ad3, inv2x);
        const g0 = runar.bbFieldAdd(hs0, at0);
        const g1 = runar.bbFieldAdd(hs1, at1);
        const g2 = runar.bbFieldAdd(hs2, at2);
        const g3 = runar.bbFieldAdd(hs3, at3);
        runar.assert(g0 == eg0);
        runar.assert(g1 == eg1);
        runar.assert(g2 == eg2);
        runar.assert(g3 == eg3);
    }
};
